http-client/src/main/java/io/micronaut/http/client/MatchingFilterResolver.java

```diff
@@ -0,0 +1,171 @@
+import java.net.URI;
+import java.util.*;
+import java.util.function.BooleanSupplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Resolver that responsible to finding all the http client filters that should be applied to given request.
+ *
+ * @since 1.2.4
+ * @author svishnyakoff
+ */
+@Primary
+@Singleton
+public class MatchingFilterResolver implements FilterResolver {
+    private static final Logger LOG = LoggerFactory.getLogger(MatchingFilterResolver.class);
+
+    private final Collection<HttpClientFilter> filters;
+    private final AnnotationMetadata annotationMetadata;
+    private final AnnotationMetadataResolver annotationMetadataResolver;
+    private final Set<String> clientIdentifiers;
+
+    /**
+     * @param annotationMetadata         annotation metadata of a declarative http client
+     * @param clientIdentifiers          client identifiers associated with a client
+     * @param annotationMetadataResolver annotation medata resolver
+     * @param filters                    all filters
+     */
+    @Inject
+    MatchingFilterResolver(@Nullable @Parameter AnnotationMetadata annotationMetadata,
+                           @Nullable @Parameter Set<String> clientIdentifiers,
+                           @Nullable AnnotationMetadataResolver annotationMetadataResolver,
+                           Collection<HttpClientFilter> filters) {
+        this.filters = filters;
+        this.annotationMetadata = Optional.ofNullable(annotationMetadata).orElse(AnnotationMetadata.EMPTY_METADATA);
+        this.annotationMetadataResolver = Optional.ofNullable(annotationMetadataResolver)
+                .orElse(AnnotationMetadataResolver.DEFAULT);
+        this.clientIdentifiers = Optional.ofNullable(clientIdentifiers).orElse(Collections.emptySet());
+    }
+
+    /**
+     * @param request    http request
+     * @param requestURI path URI
+     * @return filters that should be applied to request
+     */
+    public List<HttpClientFilter> resolveFilters(io.micronaut.http.HttpRequest<?> request,
+                                          URI requestURI) {
+        List<HttpClientFilter> filterList = new ArrayList<>();
+        String requestPath = StringUtils.prependUri("/", requestURI.getPath());
+        io.micronaut.http.HttpMethod method = request.getMethod();
+        for (HttpClientFilter filter : filters) {
+            if (filter instanceof Toggleable && !((Toggleable) filter).isEnabled()) {
+                continue;
+            }
+
+            Class<? extends Annotation>[] clientDeclaredFilterAnnotations = (Class<? extends Annotation>[]) annotationMetadata
+                    .getAnnotationTypesByStereotype(FilterAnnotation.class)
+                    .stream()
+                    .toArray(Class[]::new);
+
+            Class<? extends Annotation>[] declaredFilterAnnotations = (Class<? extends Annotation>[]) annotationMetadataResolver.resolveMetadata(filter)
```

Comment:
I don't understand why `clientDeclaredFilterAnnotations` and `declaredFilterAnnotations` are transformed into arrays here and then later transformed back into `List` objects in each `filter*` method. That will result in poor performance as you are reconstructed collections that don't need to be reconstructed.

Reply:
The comment is relevant only to `clientDeclaredFilterAnnotations`  which indeed is only used in `filter***` method so it is true - no sense to build an array that will later be converted to set.

Regarding `declaredFilterAnnotations` it still should be array cause it is passes as a paramtr to `FilterProperties`